### ISSUE TITLE
Allow for different authors between documents

### ISSUE THREAD

**Issue by bmaclach** [JacquesCarette/Drasil#1804]
Currently we have a single `authors` field in the `SystemInformation` which is used by both the SRS and code generator when printing the authors. It should instead be possible to have different authors for the code than the SRS (and different authors for potential future documents we generate).

This issue is based on https://github.com/JacquesCarette/Drasil/pull/1793#issuecomment-518218661.

**Comment by muhammadaliog3** [comment]
**Potential solution**
Change the system information data type, specified in drasil-database, 
```haskell
SI :: (....HasName c,....) => 
  {....
  , _authors :: [c]
  ,....}
```
to 
```haskell
SI :: (....HasName c,....) => 
  {....
  , _srsAuthors :: [c]
  , _codeAuthors :: [c]
  ,....}
```
There are two potential drawbacks to this solution 
1. One must explicitly state who both the authors even if they are the same person. This means that every example will have to change.
2. Say that at some point in the future of Drasil's Drasil generates another artifact that it currently doesn't, then again we will have to change the system information data type and change every example to explicitly state who the author of that document is.

This change would further propagate to the following modules (that I know of so far):

All drasil examples: change the authors fields
CodeSpec.hs: this is the only change that would be needed for all code generated.
DocumentLanguage.hs: only the mkDoc function would need to be changed.
SRS.hs: I found the following:
```haskell
data Document = Document Title Author [Section]
doc, doc' :: NamedIdea c => c -> Sentence -> [Section] -> Document
doc  sys = Document (Doc.srs `forTT` sys)
-- | Uses plural of system for title.
doc' sys = Document (Doc.srs `forTT'` sys)
```
But the weird thing is that both ```doc``` and ```doc'``` are not used anywhere in drasil (I was a able to make all the examples even after I deleted these 2 functions).

Alternative solutions: Implement the authors as a map with pairs of (Sentence, List of authors), so for example (S "SRS Author", [Alex,olu,..]). I think that this would be hard to implement for drasil-examples.
Potential Enhancement: Make drasil automatically fill in the empty author fields. However this may be difficult because the si are generated explicitly in drasil-examples and not through some constructor. So we would need to decide on a place for a function that is going to update empty author fields in the System Information.

**Comment by JacquesCarette** [comment]
Third, better solution: 'authorship' is not a global concept, but something more local, i.e. something attached to the artifacts that make up the system.

So a proper fix to this requires us to first understand, in the context of Drasil, what are the pieces that are still human-authored where credit should be given. Those pieces should then have (optional) author fields.

For now, it makes sense to push 'author' down to at least SRS and code; note that it might still make sense to have the `SI` have an author field, as authoring the overall system architecture is a separate act of authorship.

**Comment by muhammadaliog3** [comment]
Because the SRS and code generator will no longer use the authors field in the system description, they will each have to implement there own way of authorship (hence less reuse will happen). 

Summary of SRS authorship: Authorship is taken from the system information and stored directly beside the completed document. Here is the specific code if anyone wants to see:
In particular the SRS has the following datatype
```haskell 
data Document = Document Title Author [Section] 
```
and can be constructed with the following
```haskell
mkDoc :: SRSDecl -> (IdeaDict -> IdeaDict -> Sentence) -> SystemInformation -> Document
mkDoc dd comb si@SI {_sys = sys, _kind = kind, _authors = authors} =
  Document (nw kind `comb` nw sys) (foldlList Comma List $ map (S . name) authors) $
  mkSections (fillTraceMaps l (fillReqs l si)) l where
    l = mkDocDesc si dd
```
Potential solution for SRS: 
1. the mkDoc function should take in an argument of type "Author" (a type synonym for String), and make that the Author field. 

2. We could create a new "SRSinformation" datatype that stores information relevant to SRS but not other artifacts. We could potentially make SRSinformation supersede systemInformation,  as in the user cannot fill in systemInformation directly and instead must fill SRSinformation which will fill systemInformation.

Summary for Code generator: GOOL internally uses CodeSpec, which is created from the systemInformation. I will just display the function declaration to save space. 
```haskell
codeSpec :: SystemInformation -> Choices -> [Mod] -> CodeSpec
```

Potential Solution:
1. Make codeSpec explicitly take a, Author parameter instead of takingfrom the SystemInformation. 

2. Because we previously decided that Authorship is potentially an optional feature, which makes a lot of sense for code generation, authors should be a field in the Choices of Gool.

**Comment by JacquesCarette** [comment]
No, `mkDoc` should not take an Author argument. The author information should be in `SRSDecl`, which is the 'declaration of the SRS; if there is no author information there, it could be taken from `SystemInformation`, though I'd rather not.

We might need to rethink parts of `SystemInformation` too. It was never really designed, it just grew. 

For `codeSpec`, we might need `SystemInformation`, `Choices` (which could be called `DesignChoices` too), `[Mod]` **and** `CodeMetaData` (or something). Authorship could come from there.

**Comment by balacij** [comment]
[@JacquesCarette's comment](https://github.com/JacquesCarette/Drasil/issues/1804#issuecomment-643615979) explains the ideal path to resolution perfectly. Authorship is very much a 'local' idea.

So, let's say we have an SRS we're building: 'BeamBending' (my project for CAS 741). @smiths helped me out a ton with getting the theories going. He also lent me a book which explained the theory behind beam deflection analysis. The theories that I encoded in Drasil are (in my opinion) 'faithful' representations of the books derivations and theory decompositions with a Drasil-spin on them. I also took edits from @samm82 and @smiths. So, am I really the author of this? Not really -- I'm just a _transcriber_ here :smile:, @smiths and @samm82 are the editors, the _type_ of authorship is a kind of _transcription of original work_ (e.g., transcription of someone else's work in the Drasil style), and the authors of the book are the true 'authors' of the work (in some sense, I'm only an editor too!).

Ok, now let's take another example: IMs from any of our case studies. While TMs have a fair chance of always being 'transcriptions of orginal works', IMs should almost always be _ours_ because they are instances of the GDs/TMs specialized to the SRS. Thus, we are the clear authors. The traceability of 'inspiration' also appears here by dependency/refinement from the GDs/TMs.

Finally, the generated artifacts. They're generated. However, the generation step has some kind of authorship related to it too. The generator is defined by one group of people, the input information (likely the SRS) has some kind of authorship, and the choices (as @smiths explained) also has some kind of authorship which may be disjoint to the inputs. In my head, the generated artifacts should be stamped with information about the Drasil team, the SRS authors, and those who got the code generating for the SRS. This (as mentioned above) should be enough to clear the 'action items' for this ticket.

However, for the earlier 2 examples, they're quite a lot more complicated and should probably be moved to their own ticket for further discussion.

**Comment by smiths** [comment]
Interesting discussion of authorship @balacij.  For the author of a document like the SRS we should follow the normal conventions for identifying authors.  If someone contributes enough for coauthorship they should be added as a coauthor.  In the past, I've added myself as an author to some SRS documents because I invested a nontrivial amount of effort into the document.  I don't think your SRS document crosses that threshold.  @samm82 and @smiths gave you a review and some advice.  This is the same thing that often happens with journal papers, and even more frequently with books; the author asks for reviews from colleagues.  The reviewers don't generally become authors.  This discussion does bring in a new idea though.  We could add an acknowledgement entry to our database of information.  In cases where contributions are made by others, these contributions can then be acknowledged.  The SRS template does not have this information in the template, but it would be easy to add.  We'll want this information when we tackle generating journal papers.  They almost always have an acknowledgements section.  (As an aside, for journal papers we'll also want to know who funded the research.)

For the example of using the work of others to guide the presentation of information in a document (like the books you consulted for your SRS), we should handle this with citations.  The authors of the external resources do not become coauthors; they become cited authors.  Theoretical models (and other chunks) have a source field that is ideal for giving the citation information.

I don't think we want to confuse things by having the writers of the Drasil generators as coauthors of any artifacts.  People don't include the creators of any of the tools they use as coauthors.  We don't include Donald Knuth or Leslie Lamport as coauthors, even though we use LaTeX.  We don't include the developers of Haskell as coauthors.  We should acknowledge the use of Drasil as the tool we used, but we shouldn't have to do more than this.  A note in the generated documents that they were generated by Drasil, with a pointer to Drasil, should be enough.

**Comment by balacij** [comment]
Adding acknowledgements certainly sounds like a good idea :smile: 

Regarding the example of theoretical models having a source field, I think any chunk we encode could have a citation too. When we have our own homebrew chunks, they would be 'original' content where authorship is internal and local, but when we transcribe, for example, encodings, such as WebAssembly, Makefile, JSON, etc., we could have a description attached to the chunk that says some person (an author) transcribed the work of some others (the source). Another example: if we had a `SRS` data type, then its citation would probably be your SRS paper (@smiths).

Attaching information to chunks this way would probably allow us to automate authorship of an SRS. For example, for an SRS to be 'conventional', its authorship should include anyone who implemented any nonzero number of 'problem-specific chunks' (such as IMs, PSD, GSD, derivations, etc.).

**Comment by JacquesCarette** [comment]
We don't want to go overboard on this: content has to be substantial enough to claim some level of authorship. [Contra that opinion: git itself lets tiny PRs have authorship associated with it.]

Certainly we want larger pieces to have their own separate authorship notion. We can slowly push that 'down' onto smaller units as it becomes necessary.
